Taproot: integrate btcec/v2 and btcwallet changes to support Taproot key spend paths in lnd's wallet

guggero opened the pull request:
End-to-end integration of all the changes required to get the Taproot functionality into `lnd`.

Depends on ~https://github.com/btcsuite/btcd/pull/1787~, https://github.com/btcsuite/btcwallet/pull/792 and ~https://github.com/lightningnetwork/lnd/pull/6285~.

Fixes https://github.com/lightningnetwork/lnd/issues/6266.
Fixes https://github.com/lightningnetwork/lnd/issues/6267.
Fixes https://github.com/lightningnetwork/lnd/issues/6330.

guggero commented on lnwire/signature_test.go:
```diff
@@ -24,67 +25,80 @@ func TestSignatureSerializeDeserialize(t *testing.T) {
+	//sScalar.SetByteSlice(s.Bytes())
+	//sig = ecdsa.NewSignature(rScalar, sScalar)
+	//err = signatureSerializeDeserialize(sig)
+	//if err.Error() != "Pre/post-serialize Ss don't match: 115792089237316"+
```
Something seems to be serialized differently here. @Roasbeef, not sure if I did anything wrong here?

guggero replied:
I wasn't able to re-create this test case fully, at least not in regard to its error message. The problem is: the sigs don't match (as expected) because `Serialize()` correctly fixes/unifies the low-s value. But we can't access the internal state to show the difference, we can only call `Serialize()` to give us the raw bytes (which again fix the low-s value).
So tl;dr is: we can only assert that the sig is different but cannot show the actual difference of its internal state (unless we do some nasty reflection magic.

Roasbeef replied:
Ah makes sense, the `Serialize()` we use `Serialize()` we use is the `dcrec` version that does indeed fix those issues as part of encoding: https://github.com/decred/dcrd/blob/dcrec/secp256k1/v4.0.1/dcrec/secp256k1/ecdsa/signature.gao#L89-L96. The prior one does as well, but this one also attempts to more aggressively strip out leading zeros. 

Nice to tie up this loose end to make sure it doesn't bit us in the future!

guggero commented on lnwire/signature_test.go:
```diff
@@ -24,67 +25,80 @@ func TestSignatureSerializeDeserialize(t *testing.T) {
-		t.Fatalf("R = 2N, S = 128, R should be over 32 bytes: %s",
-			err.Error())
-	}
+	//r = new(big.Int).Set(btcec.S256().N)
```
We can't overflow anymore because we're using ModNScalar. Should we keep this test case and try to replicate it in a different way? Or just add a test case for _de-serializing_ an invalid value?

Roasbeef replied:
If you pass something that's too large, then `SetByteSlice` will return an error, so that can be checked.

guggero replied:
You mean the `overflow` boolean? Yeah, I added that as a check indeed.

guggero commented on contractcourt/breacharbiter.go:
```diff
@@ -1441,7 +1441,7 @@ func (b *BreachArbiter) sweepSpendableOutputsTxn(txWeight int64,
 	// Create a sighash cache to improve the performance of hashing and
 	// signing SigHashAll inputs.
-	hashCache := txscript.NewTxSigHashes(txn)
+	hashCache := txscript.NewTxSigHashesV0Only(txn)
```
@Roasbeef: This whole commit makes me a bit uneasy. And we'll probably have to add way more logic. Can't we somehow simplify this in `btcd` itself? Since this seems to only be needed for SigHashAnyoneCanPay and we never do that in `lnd` anyway?

Roasbeef replied:
This `NewTxSigHashesV0Only` function doesn't exist rn right? What do you mean it makes you uneasy? If we specify the `CannedPrevOutFetcher` (or w/e it's called) here, we make sure we're always doing the right thing. In each instance, we also already know what the prev out is.

guggero replied:
I moved the `NewTxSigHashesV0Only` from my `btcd` branch into `lnd` and updated most of the non-test code to not actually use it (and set up a proper prev output fetcher instead).
Turned out we do have all the info about all the spent outputs in pretty much all cases. which makes me feel way better about this. The `NewTxSigHashesV0Only ` is only used in unit test code or cases where we _know_ we currently only have v0 inputs (e.g. a channel funding output).

guggero replied:
Oops, yeah, I forgot to update that commit. Fixed.

Roasbeef commented on lnwire/signature_test.go:
```diff
@@ -24,67 +25,80 @@ func TestSignatureSerializeDeserialize(t *testing.T) {
 	// Check R = N-1, S = N>>1.
-	sig.S.Set(btcec.S256().N)
-	sig.S.Rsh(sig.S, 1)
+	s := new(big.Int).Set(btcec.S256().N)
```
Why mix big ints, and the mod n scalar here?

guggero commented:
I extracted most of the diff into https://github.com/lightningnetwork/lnd/pull/6285 which doesn't depend on any unmerged `btcd` PRs and can be merged whenever.

Roasbeef reviewed (commented):
Thought I posted these comments...

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

guggero commented on autopilot/graph.go:
```diff
@@ -17,12 +18,13 @@ import (
-	}
-	_, _ = testSig.R.SetString("63724406601629180062774974542967536251589935445068131219452686511677818569431", 10)
-	_, _ = testSig.S.SetString("18801056069249825825291287104931333862866033135609736119018462340006816851118", 10)
+	testR, _    = new(big.Int).SetString("63724406601629180062774974542967536251589935445068131219452686511677818569431", 10)
```
Right, switched over to using hex notation and plugging that into `SetByteSlice` directly.

guggero commented on go.mod:
```diff
@@ -101,6 +103,19 @@ replace github.com/ulikunitz/xz => github.com/ulikunitz/xz v0.5.8
+// gopkg.in/fsnotify.v1 that isn't go mod compatible. We fix that import error
+// by replacing ginko (which is only a test library anyway) with a more recent
+// version.
+replace github.com/onsi/ginkgo => github.com/onsi/ginkgo v1.14.2
```
Going to create a PR for that (and for https://github.com/btcsuite/btcd/issues/1822).

Roasbeef replied:
Merged that in, once we merge in the zero-ish dep `btcec/v2` module, then we can update this as well.

guggero commented on input/size.go:
```diff
@@ -519,6 +519,12 @@ const (
+	//	- NumberOfWitnessElements: 1 byte
+	//	- sigLength: 1 byte
+	//	- sig: 64 bytes
+	TaprootKeyPathWitnessSize = 1 + 1 + 64
```
Ah right. Added a second constant and used the longer one in the `SizeUpperBound` method.

guggero commented on lnrpc/lightning.proto:
```diff
@@ -1102,12 +1102,15 @@ message ListUnspentResponse {
     UNUSED_WITNESS_PUBKEY_HASH = 2;
     UNUSED_NESTED_PUBKEY_HASH = 3;
+    TAPROOT_PUBKEY = 4;
+    UNUSED_TAPROOT_PUBKEY = 5;
```
Yeah, I don't think this is used very often but we allow to ask the wallet for the last "unused" address of a type. And it will return the same address until funds were detected on it (and the address is internally marked as "used").

guggero commented on input/signdescriptor.go:
```diff
@@ -66,6 +66,8 @@ type SignDescriptor struct {
 	// otherwise an invalid signature may be generated.
 	Output *wire.TxOut
 
+	Utxos []*wire.TxOut
```
Yes, though I never fully implemented that part. Have a full version up now though, using a `txscript.PrevOutputFetcher` in the sign descriptor.

guggero commented:
Rebased and addressed all current comments. Also made the itest a bit less confusing and bit more explicit.

Not sure if the new `AssembleTapscriptSpendWitness` RPC should be added in this PR or a follow-up one.

guggero commented on lnwallet/btcwallet/signer.go:
```diff
@@ -379,6 +381,45 @@ func (b *BtcWallet) SignOutputRaw(tx *wire.MsgTx,
 
+	if txscript.IsPayToTaproot(signDesc.Output.PkScript) {
+		trAddr, err := btcutil.NewAddressTaproot(
+			signDesc.Output.PkScript[2:], b.netParams,
```
Fixed by doing things in another way.

Roasbeef commented on lntest/itest/lnd_taproot_test.go:
```diff
@@ -80,4 +89,153 @@ func testTaproot(net *lntest.NetworkHarness, t *harnessTest) {
+		PkScript: p2trPkScript,
+		Value:    800_000,
+	}}
+	signResp, err := net.Alice.SignerClient.SignOutputRaw(
```
I don't see where in this diff we allow the user to specify a _schnorr_ signature here?

Roasbeef commented on lntest/itest/lnd_taproot_test.go:
```diff
@@ -80,4 +89,153 @@ func testTaproot(net *lntest.NetworkHarness, t *harnessTest) {
+			TxHex: buf.Bytes(),
+		},
+	)
+	require.NoError(t.t, err)
```
Should also mine a block to assert inclusion of the spend.

Roasbeef commented on lnwallet/btcwallet/signer.go:
```diff
@@ -368,6 +369,33 @@ func (b *BtcWallet) SignOutputRaw(tx *wire.MsgTx,
 
+	// In case of a taproot output any signature is always a Schnorr
+	// signature, based on the new tapscript sighash algorithm.
+	if txscript.IsPayToTaproot(signDesc.Output.PkScript) {
```
Ah ok, we'll do it depending on the output....def easier UX wise as well since the user can just send in all their inputs. 

Doesn't necessarily need to be a part of this PR (maybe will be added re musig2)

Roasbeef replied:
With the way things are right now, it _always_ assumes the tapscript sighash, so a user wouldn't be able to generate just normal keyspend signatures here. 

So either way, looks like we'll need _some_ additional arg/enum here to reflect all the possibilities.

guggero replied:
Ah, you're right, we're missing things to actually support signing with MuSig. I'll think a bit about that. Might make sense to do it in a separate PR (that also pulls in the MuSig library, which this PR does not yet do).

Roasbeef replied:
Yeah totally agree we should do the musig stuff in another PR.

Roasbeef commented on lnwallet/btcwallet/signer_test.go:
```diff
@@ -167,6 +173,25 @@ func TestBip32KeyDerivation(t *testing.T) {
+	)
+	require.NoError(t, err)
+
+	// mgr.ImportWitnessScript()
```
Leaving a comment here so this doesn't get lost.

guggero replied:
Thanks, added a TODO comment and split into its own commit.

Roasbeef reviewed (commented):
Great to see this PR continue to progress! 

I've done another round of review, and noted some other additions we may want to make in the RPC interface, and also a scenario where the current implicit ecdsa vs schnorr determination in `SignOutputRaw` would generate an invalid signature (user wants to do keyspend but we do a script path spend).

guggero commented on input/signdescriptor.go:
```diff
@@ -74,6 +74,11 @@ type SignDescriptor struct {
+	// PrevOutputFetcher is an interface that can return the output
+	// information on all UTXOs that are being spent in this transaction.
+	// This MUST be set when spending Taproot outputs.
+	PrevOutputFetcher txscript.PrevOutputFetcher
```
`txscript.PrevOutputFetcher` is an interface, so it's basically a pointer. Or you mean a pointer to the interface to avoid that nil equality check against interfaces?

guggero commented on lnwallet/chanfunding/wallet_assembler.go:
```diff
@@ -374,3 +380,55 @@ func (w *WalletAssembler) FundingTxAvailable() {}
+//
+// TODO(guggero): Support dual funding with p2tr inputs, currently only segwit
+// v0 inputs are supported.
+type SegWitV0DualFundingPrevOutputFetcher struct {
```
For the OG dual funding tests in the wallet integration tests.

guggero commented on lnwallet/interface.go:
```diff
@@ -161,6 +161,11 @@ type WalletController interface {
 	// a non-nil error value of ErrNotMine should be returned instead.
 	FetchInputInfo(prevOut *wire.OutPoint) (*Utxo, error)
 
+	// FetchPrevOutput attempts to fetch the previous output referenced by
```
Ah yes, good idea. Done.

guggero commented on lnwallet/btcwallet/signer.go:
```diff
@@ -391,6 +392,40 @@ func (b *BtcWallet) SignOutputRaw(tx *wire.MsgTx,
 func (b *BtcWallet) ComputeInputScript(tx *wire.MsgTx,
 	signDesc *input.SignDescriptor) (*input.Script, error) {
 
+	if txscript.IsPayToTaproot(signDesc.Output.PkScript) {
```
Good idea! Done.

dstadulis commented:
We're building several other PRs off of this one

action items:
continue to feedback requirements from lnd which themselves have requirements from LL products

guggero commented on chainntnfs/txnotifier.go:
```diff
@@ -322,6 +335,20 @@ func NewSpendRequest(op *wire.OutPoint, pkScript []byte) (SpendRequest, error) {
+		// zero Taproot key that we'll compare this spend request to.
+		r.PkScript = ZeroTaprootPkScript
+	}
+
```
No, fortunately the `ConfRequest` is the easy case, since we only care about a certain _output_ being confirmed. So if you look at `ConfRequest.MatchesTx()`, we directly compare the `PkScript` of the transaction's _outputs_ with the confirmation request and don't do any `txscript.ComputePkScript()` magic that we do in `SpendRequest.MatchesTx()`.

guggero commented:
Changes since last comment:
 - Confirmation and spend requests for Taproot outputs now fully work, with the caveat that the outpoint must always be specified.
 - Added weight estimation for Taproot key spend and script spend inputs
 - Added helper functions for assembling the control block
 - Added unit tests for Tapscript address import into the wallet
 - Added Taproot key spend (BIP-0086) addresses to the on-chain recovery itest

With that, all remaining TODOs that I can think of are done, so taking this PR out of draft mode.

Roasbeef commented on input/signdescriptor.go:
```diff
@@ -61,6 +61,11 @@ type SignDescriptor struct {
+	// TaprootKeySpend indicates that instead of a witness script being
+	// spent by the signature that results from this signing request, a
+	// taproot key spend is performed instead.
+	TaprootKeySpend bool
```
Assuming we'd add on the additional context here re the leaf being revealed, etc -- I would think that would be enough to go off on (still getting through the latest diff).

Roasbeef replied:
Ah ok, revisiting the prior commit, I see that we handle the scrip path based on a combo of this bool and the output type. I guess ultimately this single bool is better than exposing 4+ enums that the user need to juggle with.

Roasbeef commented on lntest/itest/lnd_taproot_test.go:
```diff
@@ -272,3 +273,150 @@ func testTaprootScriptSpend(ctxt context.Context, t *harnessTest,
+	internalKey, err := btcec.ParsePubKey(keyDesc.RawKeyBytes)
+	require.NoError(t.t, err)
+
+	// Let's create a taproot script output now. This is a hash lock with a
```
Not a huge blocker, but this fragment shows up a few times in the integration tests and could be moved to a simple routine.

Roasbeef commented on chainntnfs/txnotifier.go:
```diff
@@ -322,6 +335,20 @@ func NewSpendRequest(op *wire.OutPoint, pkScript []byte) (SpendRequest, error) {
 
+	// For Taproot spends we have the main problem that for the key spend
+	// path we cannot derive the pkScript from only looking at the input's
+	// witness. So we need to rely on the outpoint information alone.
```
We should add a TODO here to eventually be able to detect classes of scriptpath spends, since we can re-compute the previous output script from the control block.

guggero replied:
Good idea. Added the TODO.

Roasbeef reviewed (commented):
Stoked to see this PR leave draft mode! Excellent set of changes that cross several areas of the codebase. No major comments this time around. 

I'll be running this PR now on mainnet+testnet to give it a spin and see if I detect any potential improvements to logging, or error messages, etc.

guggero commented on go.sum:
```diff
@@ -547,6 +547,27 @@ github.com/lightningnetwork/lightning-onion v1.0.2-0.20210520211913-522b799e65b1
 github.com/lightningnetwork/lightning-onion v1.0.2-0.20220211021909-bb84a1ccb0c5 h1:TkKwqFcQTGYoI+VEqyxA8rxpCin8qDaYX0AfVRinT3k=
 github.com/lightningnetwork/lightning-onion v1.0.2-0.20220211021909-bb84a1ccb0c5/go.mod h1:7dDx73ApjEZA0kcknI799m2O5kkpfg4/gr7N092ojNo=
 github.com/lightningnetwork/lnd v0.14.2-beta/go.mod h1:BKTR+jbfcyFwsOPb3m8HaM09YmZF/SsbWd5UTCgDZlo=
+github.com/lightningnetwork/lnd/cert v1.1.0/go.mod h1:3MWXVLLPI0Mg0XETm9fT4N9Vyy/8qQLmaM5589bEggM=
```
Ah, you're right. Good catch! We're pulling in the complete `lnd@v0.14.2-beta` from `healthcheck`. Created https://github.com/lightningnetwork/lnd/pull/6350 to fix that.

guggero commented on contractcourt/breacharbiter_test.go:
```diff
@@ -1377,7 +1377,8 @@ func getSpendTransactions(signer input.Signer, chanPoint *wire.OutPoint,
 
-	hashCache := txscript.NewTxSigHashes(htlcSweep)
+	cannedFetcher := txscript.NewCannedPrevOutputFetcher(nil, 0)
+	hashCache := txscript.NewTxSigHashes(htlcSweep, cannedFetcher)
```
No reason. I think I wanted to make this work properly but then figured we didn't have all the info we needed and reverted it.

guggero replied:
Reverted this back to `NewTxSigHashesV0Only`.

positiveblue reviewed (approved):
LGTM, I think we are getting close to have it ready to be merged :tada:. Release notes are missing but functionality wise I think it is complete.

guggero commented:
Added release notes.

guggero commented:
Fixed itests too. Ready to be merged IMO. How did your mainnet tests come along, @Roasbeef?

Roasbeef reviewed (approved):
LGTM 🐞

Roasbeef commented on lntest/itest/lnd_taproot_test.go:
```diff
@@ -257,6 +259,147 @@ func testTaprootScriptSpend(ctxt context.Context, t *harnessTest,
+				Output:          utxoInfo[0],
+				InputIndex:      0,
+				KeyDesc:         keyDesc,
+				SingleTweak:     dummyKeyTweak[:],
```
Does this actually need to be set?

Roasbeef replied:
Ah ok it's just making sure that combination work as well. Though I don't think it would ever really be used....

Roasbeef commented:
Thx for the report @Sjors, I converted you comment into an issue so we can track it properly
